chem_tensorflow_sparse.py

```diff
@@ -122,33 +120,73 @@ def compute_final_node_representations(self) -> tf.Tensor:
                 node_states_per_layer.append(node_states_per_layer[-1])
                 for step in range(num_timesteps):
                     with tf.variable_scope('timestep_%i' % step):
-                        incoming_messages = []  # list of v x D
+                        # Used shape abbreviations:
+                        #   V ~ number of nodes
+                        #   D ~ state dimension
+                        #   E ~ number of edges of current type
+                        #   M ~ number of messages (sum of all E)
+                        messages = []  # list of tensors of messages of shape [E, D]
+                        message_source_states = []  # list of tensors of edge source states of shape [E, D]
+                        message_targets = []  # list of tensors of message targets of shape [E]
+                        message_edge_types = []  # list of tensors of edge type of shape [E]
 
                         # Collect incoming messages per edge type
-                        for adjacency_matrix in adjacency_matrices:
-                            incoming_messages_per_type = tf.sparse_tensor_dense_matmul(adjacency_matrix,
-                                                                                       node_states_per_layer[-1])  # v x D
-                            incoming_messages.extend([incoming_messages_per_type])
-
-                        # Pass incoming messages through linear layer:
-                        incoming_messages = tf.concat(incoming_messages, axis=1)  # v x [2 *] edge_types
-                        messages_passed = tf.matmul(incoming_messages,
-                                                    self.weights['edge_weights'][layer_idx])  # v x D
+                        for edge_type_idx, adjacency_list_for_edge_type in enumerate(self.placeholders['adjacency_lists']):
+                            edge_sources, edge_targets = adjacency_list_for_edge_type[:, 0], adjacency_list_for_edge_type[:, 1]
+                            edge_source_states = tf.nn.embedding_lookup(params=node_states_per_layer[-1],
+                                                                        ids=edge_sources)  # Shape [E, D]
+                            all_messages_for_edge_type = tf.matmul(edge_source_states,
+                                                                   self.weights['edge_weights'][layer_idx][edge_type_idx])  # Shape [E, D]
+                            messages.append(all_messages_for_edge_type)
+                            message_source_states.append(edge_source_states)
+                            message_targets.append(edge_targets)
+                            message_edge_types.append(tf.ones_like(edge_targets, dtype=tf.int32) * edge_type_idx)
```

Comment:
This (and related operations) can be done once outside of the loop, since the edges don't change across steps. Might speedup things?

Reply:
... and similarly fo `edge_sources` and `edge_targets`

Reply:
Done.